README


This archive contains the distribution Params-Validate,
version 1.12:

  Validate method/function parameters

This software is Copyright (c) 2014 by Dave Rolsky and Ilya Martynov.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)


This README file was generated by Dist::Zilla::Plugin::Readme v5.019.



Changes
1.16    2015-01-07

- The changes in 1.14 introduced a memory leak any time a callback was called
  and it did not throw an error that was a reference. This affected the
  DateTime constructor and probably many other things. Reported by David
  Kayal. RT #101380.


1.15    2015-01-01

- No changes from 1.14


1.14    2014-12-20 (TRIAL RELEASE)

- Callbacks can now die to provide a custom error message or exception
  object. Requested by multiple people. Addresses RT #95701 and will allow
  MooseX::Params::Validate to use messages provided by the type object.


1.13    2014-06-28

- Fix my brain damage so that this code compiles with Perl 5.14.


1.12    2014-06-27

- Temporarily skip tests that use Readonly with Readonly 1.50+ until various
  bugs are worked out in Readonly.


1.11    2014-06-26

- Fixes for MSVC compilation. Patch by J.R. Mash. PR #1.


1.10    2014-05-11

- Require Readonly 1.03+ and Scalar::Util 1.20 for testing with
  Readonly. AFAICT, earlier versions of Readonly worked very differently,
  and/or this is related to Scalar::Util. Either way, this test isn't that
  important anyway. RT #95402.


1.09    2014-05-04

- A call to validate() where the spec was a reference to a Readonly hash would
  segfault. Reported by Salvatore Bonaccorso. RT #80572.

- When using the XS implementation, attempting to validate parameters with
  tainted values would cause an "Insecure dependency in eval_sv() while
  running with -T switch" exception from Perl *if the validation failed*. This
  is fixed, but only for Perl 5.14+. Older Perls do not include the necessary
  XS API. RT #70174.

- Fixed some potential segfaults from the XS implementation including the one
  reported by Анатолий Гришаев in RT #50412. However, the XS code in general
  is pretty crufty and tends to assume that values which could be NULL aren't.

- Fixed bad example of state() in POD. Reported by Salvatore Bonaccorso. RT
  #80250.


1.08    2013-06-07

- The handling of defaults in the XS version of validate_pos was broken. The
  default were simply pushed into the returned array, rather than being
  explicitly stored in the right index based on the order of the specs passed
  to validate_pos(). RT #83780.


1.07    2012-10-26

- Params::Validate's XS implementation would overwrite $@ when validation subs
  were called. Patch by Salvador Fandino. RT #80124.

- The use of state in the POD was broken. It only works with scalars. Reported
  by Salvatore Bonaccorso. RT #80250.


1.06    2012-02-10

- Shut up warnings when XS is loaded on older 5.8.x Perl versions. Reported by
  Aaron James Trevena. RT #74742.


1.05    2012-02-08

- The XS code had a code path where it could pass the contents of a Perl
  variable as the first argument to the XS croak() subroutine. This subroutine
  is like printf(), and should receive a format string as its first
  argument. According to RT #74777, this can lead to segfaults on some systems.

  This could in theory be a security bug, but it's very unlikely that
  untrusted user input could end up being passed to this croak(). It is called
  when a spec specifies a "depend" value on another parameter. The value of
  the "depend" parameter was passed in the first argument to croak().

  Reported by Andreas Voegele.


1.04    2012-02-08

- Use the latest Module::XSOrPP dzil plugin to generate a saner Build.PL. No
  need update if you're using an earlier version.


1.03    2012-02-06

- This release uses Module::Implementation to handle loading the XS or pure
  Perl implementation of Params::Validate.


1.02    2012-02-06

- The previous release never loaded the XS implementation, even if it had been
  compiled.

- With newer versions of Perl, the pure Perl implementation treated regexp
  objects differently than the XS implementation. They should be treated as
  belonging to the SCALARREF type for backwards compatibility.

- These two bugs combined managed to break the test suites of a number of
  modules on CPAN. This release should fix them.


1.01    2012-02-05

- The generated Build.PL detects whether or not the machine it's run on has a
  working compiler, and disables XS automatically if it doesn't. Fixes RT
  #44719.


1.00    2011-06-11

- Fix documentation indexing on search.cpan.org. Reported by Lutz Gehlen. RT
  #68756.


0.99    2011-05-27

- The metadata file - META.{json,yml} - had references to a different distro
  because of a copy and paste error. Reported by Bernhad Graf. RT #68514.


0.98    2011-04-23

- Attribute::Params::Validate's $VERSION regressed in 0.96. It now manually
  set to 1.07. Reported by tokuhirom. RT #67715.


0.97    2011-04-19

- The spec validation added in 0.96 broke MooseX::Params::Validate, so it has
  been removed until I can work out how to make the two modules play nice.


0.96    2011-04-19

- The XS version of the code always called Carp::confess, regardless of
  whether you provided your own on_fail callback. Reported by Scott Bolte. RT
  #66359.

- There were a couple spots that called eval without localizing $@ first. RT
  #58087.

- The parameters for each key validation (can, isa, regex) are now checked,
  and an error is thrown if any of the keys are not valid. Basically, we
  validate the validation spec. Based on a patch by Andreas Faafeng. RT
  #57831.

- Lots of little pod fixes. RT #61002.

- This module now requires Perl 5.8.1.

- Switched to version 2.0 of the Artistic License.


0.95    2010-02-16

- Skip t/29-taint-mode.t entirely, since it seems to randomly fail for people,
  despite already being marked a todo test.


0.94    2009-12-01

- Removed the Makefile.PL entirely. If it was used then the XS code was never
  built, because I'm using the Module::Build XS layout. Reported by jawnsy on
  IRC.


0.93    2009-11-30

- If the pure Perl version of validate() received a hash reference it would
  alter that reference directly. Reported by Paul Eckhardt. RT #51155.


Build.PL

# This file was automatically generated by inc::MyModuleBuild v(dev).
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "c_source" => "c",
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Validate method/function parameters",
  "dist_author" => [
    "Dave Rolsky <autarch\@urth.org>",
    "Ilya Martynov <ilya\@martynov.org>"
  ],
  "dist_name" => "Params-Validate",
  "dist_version" => "1.13",
  "license" => "artistic_2",
  "module_name" => "Params::Validate",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Attribute::Handlers" => "0.79",
    "Carp" => 0,
    "Exporter" => 0,
    "Module::Implementation" => 0,
    "Scalar::Util" => "1.10",
    "XSLoader" => 0,
    "attributes" => 0,
    "perl" => "5.008001",
    "strict" => 0,
    "vars" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Devel::Peek" => 0,
    "File::Spec" => 0,
    "File::Temp" => 0,
    "Test::Fatal" => 0,
    "Test::More" => "0.88",
    "Test::Requires" => 0,
    "Tie::Array" => 0,
    "Tie::Hash" => 0,
    "base" => 0,
    "lib" => 0,
    "overload" => 0
  }
);


my %fallback_build_requires = (
  "Devel::Peek" => 0,
  "File::Spec" => 0,
  "File::Temp" => 0,
  "Module::Build" => "0.3601",
  "Test::Fatal" => 0,
  "Test::More" => "0.88",
  "Test::Requires" => 0,
  "Tie::Array" => 0,
  "Tie::Hash" => 0,
  "base" => 0,
  "lib" => 0,
  "overload" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

my $skip_xs;
if ( grep { $_ eq '--pp' } @ARGV ) {
    $skip_xs = 1;
}
elsif ( ! $build->have_c_compiler() ) {
    $skip_xs = 1;
}

if ($skip_xs) {
    $build->build_elements(
        [ grep { $_ ne 'xs' } @{ $build->build_elements() } ] );
}
$build->create_build_script;
